### PR TITLE
[Event Collector] - GitGuaridan

### PR DESCRIPTION
**Pull request by MLainer1**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-dc.paloaltonetworks.com/browse/CIAC-9236)

## Description
Creating a new event collector for GitGuardian

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py
Comment (dantavori) [inline Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py:1]: in both test-module and get-events, better to just call fetch_events() for better debugging coverage

**Review by dantavori — COMMENTED** [review]
looks great!
I will approve after we have an answer about the 2nd endpoint, and the demo

File: Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+    """Client class to interact with the service API
+
+    This Client implements API calls, and does not contain any Demisto logic.
+    Should only do requests and return data.
+    It inherits from BaseClient defined in CommonServer Python.
+    Most calls use _http_request() that handles proxy, SSL verification, etc.
+    For this HelloWorld implementation, no special attributes defined
+    """
+
+    def search_events(
```
Comment (dantavori) [inline Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py:37]: I'd merge this function with `fetch_events()`

File: Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+    audit_logs = audit_logs[:limit]
+
+    hr = tableToMarkdown(
+        name="Test Event - incidents",
+        t=incidents,
+        headers=["display_name", "id", "created_at", "type", "_time"],
+        removeNull=True,
+    )
+    hr += tableToMarkdown(
+        name="Test Event - audit_logs",
```
Comment (dantavori) [inline Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py:269]: test event?

File: Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+                break
+
+        last_fetched_events_ids, next_fetch_url, is_nextTrigger = self.handle_events(events, event_type, to_fetch_time, next_url, prev_run_fetched_event_ids)  # noqa: E501
+
+        if response.get("count") == 0:
+            demisto.debug("GG: No events were fetched.")
+
+        return events, last_fetched_events_ids, next_fetch_url, is_nextTrigger
+
+    def handle_events(self, events: list, event_type: str, to_fetch_time: str, next_url: str,
```
Comment (dantavori) [inline Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py:119]: please complete docstring

File: Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+        from_fetch_time = last_run.get("from_fetch_time", "")
+        to_fetch_time = last_run.get("to_fetch_time", "")
+        last_fetched_event_ids = last_run.get("last_fetched_ids", [])
+        next_url_link = last_run.get("next_url_link", "")
+
+        events, last_fetched_event_ids, next_fetch_url, is_nextTrigger = self.retrieve_events(
+            from_fetch_time, to_fetch_time, max_events_per_fetch, last_fetched_event_ids, event_type, next_url_link
+        )
+
+        if is_nextTrigger:
```
Comment (dantavori) [inline Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py:61]: `nextTrigger` is the name of the field we send to the lastRun object that indicates how many seconds to wait between the fetches.
`is_nextTrigger` at this point doesn't really mean anything, maybe:
```suggestion
        if is_pagination_in_progress:
```

File: Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+                )
+            else:
+                demisto.debug(f"GG: Fetching events using the params: {params}")
+                response = self._http_request(
+                    method="GET",
+                    url_suffix=EVENT_TYPE_TO_ENDPOINT.get(event_type, ''),
+                    params=params,
+                    retries=3,
+                )
+            new_events = self.remove_duplicated_events(response.get("results"), prev_run_fetched_event_ids)
```
Comment (dantavori) [inline Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py:105]: why would we run this on the same `prev_run_fetched_event_ids` over multiple pages?

File: Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector_test.py
Comment (dantavori) [inline Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector_test.py:1]: let's add a test for the nextTrigger thing

**Review by dantavori — COMMENTED** [review]
looks good, see notes

File: Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py (hunk 5 above)
Comment (MLainer1) [reply Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py:105]: The API returns results consequently in the incident ID, and not in the last_occurence_date field, which means that if event 1 happened after event number 2, they will still get returned as [event 1, event 2]

File: Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py (hunk 1 above)
Comment (MLainer1) [reply Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.py:29]: I prefer to keep them as separate functions

**Review by dantavori — APPROVED** [review]

File: Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.yml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  type: 9
+  required: true
+  section: Connect
+- defaultvalue: 5000
+  section: Collect
+  display: Max number of events per fetch
+  name: max_events_per_fetch
+  required: false
+  type: 0
+  additionalinfo: Max number of events per fetch per event type. It is recommended to choose a number with multiples of 1000.
```
Comment (ShirleyDenkberg) [inline Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector.yml:27]:
```suggestion
  additionalinfo: Maximum number of events per fetch per event type. It is recommended to choose a number in multiples of 1000.
```

File: Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector_description.md [hunk 7]
```diff
@@ -0,0 +1,8 @@
+Use this integration to collect events automatically from GitGuardian.
+You can also use the ***gitguardian-get-events*** command to manually collect events.
+
+## API Key
+To get the API key, go to the GitGuardian Enterprice web interface, click on the user icon in the top right corner, and select ***API***, then click ***Create API Key***.
```
Comment (ShirleyDenkberg) [inline Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector_description.md:5]:
```suggestion
To get the API key, go to the GitGuardian Enterprice web interface, click on the user icon in the top right corner, and select **API**, then click **Create API Key**.
```

File: Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector_description.md [hunk 8]
```diff
@@ -0,0 +1,8 @@
...
+The scopes needed for the key are: 
+1. incidents:read
+2. audit_log:read
```
Comment (ShirleyDenkberg) [inline Packs/GitGuardian/Integrations/GitGuardianEventCollector/GitGuardianEventCollector_description.md:8]:
```suggestion
- incidents:read
- audit_log:read
```

**Comment by ShirleyDenkberg** [comment]
@MLainer1 Need content for the pack readme.
@dantavori Doc review completed.

**Pull request merged**